src/pybreaker.py

```diff
@@ -363,6 +367,125 @@ def opened_at(self, datetime):
+            raise ImportError("CircuitRedisStorage can only be used if the required dependencies exist")
+
+        self._redis = redis_object
+        self._namespace_name = namespace
+        self._fallback_circuit_state = fallback_circuit_state
+
+        self._redis.setnx(self._namespace('fail_counter'), 0)
+        self._redis.setnx(self._namespace('state'), str(state))
+
+    @property
+    def state(self):
+        """
+        Returns the current circuit breaker state.
+        """
+        try:
+            return self._redis.get(self._namespace('state')).decode('utf-8')
+        except RedisError:
+            return self.fallback_circuit_state
+
+    @state.setter
+    def state(self, state):
+        """
+        Set the current circuit breaker state to `state`.
+        """
+        try:
+            self._redis.set(self._namespace('state'), str(state))
+        except RedisError:
+            pass
+
+    def increment_counter(self):
+        """
+        Increases the failure counter by one.
+        """
+        try:
+            self._redis.incr(self._namespace('fail_counter'))
+        except RedisError:
+            pass
+
+    def reset_counter(self):
+        """
+        Sets the failure counter to zero.
+        """
+        try:
+            self._redis.set(self._namespace('fail_counter'), 0)
+        except RedisError:
+            pass
+
+    @property
+    def counter(self):
+        """
+        Returns the current value of the failure counter.
+        """
+        try:
+            value = self._redis.get(self._namespace('fail_counter'))
+            if value:
+                return int(value)
+            else:
+                return 0
+        except RedisError:
+            return 0
+
+    @property
+    def opened_at(self):
+        """
+        Returns a datetime object of the most recent value of when the circuit
+        was opened.
+        """
+        try:
+            timestamp = self._redis.get(self._namespace('opened_at'))
+            if timestamp:
+                return datetime(*time.gmtime(int(timestamp))[:6])
+        except RedisError:
+            return None
+
+    @opened_at.setter
+    def opened_at(self, datetime):
+        """
+        Atomically sets the most recent value of when the circuit was opened
+        to `datetime`. Stored in redis as a simple integer of unix epoch time.
+        To avoid timezone issues between different systems, the passed in
+        datetime should be in UTC.
+        """
+        try:
+            key = self._namespace('opened_at')
+            def set_if_greater(pipe):
+                current_value = pipe.get(key)
+                next_value = int(calendar.timegm(datetime.timetuple()))
+                pipe.multi()
+                if not current_value or next_value > int(current_value):
+                    pipe.set(key, next_value)
+
+            self._redis.transaction(set_if_greater, key)
```

Comment:
This does work - it's all referencing the `key` variable on line 470, in the outer method scope. The `transaction` method expects a callable which it passes the `pipe` to - see the last example in this section in the redis-py docs: https://github.com/andymccurdy/redis-py#pipelines. I can see how it'd be confusing though - one alternative would be to explicitly `curry` the `set_if_greater` call. What are your thoughts?